### PR TITLE
Gregord Test CLI

### PR DESCRIPTION
**Pull request by jacobhaven**
r? @maxtaco @patrickxb

### CODE REVIEW

File: test/gregorcli/main.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+	if res != "pong" {
+		log.Fatalf(`res (%s) != "pong"`, res)
+	}
+
+	// Test ConsumeMessage
+	var of gregor1.ObjFactory
+	uid := gregor1.UID([]byte{0x10, 0x26, 0x94, 0xB4, 0x46, 0x42, 0xA0, 0x68, 0x3D, 0x32, 0x1A, 0xD3, 0xF8, 0xD7, 0xAE, 0x5F})
```
Comment (jacobhaven) [inline test/gregorcli/main.go]: This probably shouldn't be hard-coded here, but otherwise we'd need to parse the UID.

Comment (maxtaco) [inline test/gregorcli/main.go]: Why isn't it necessary to authenticate here?

Comment (jacobhaven) [inline test/gregorcli/main.go]: You're looking at an out-of-date diff

File: test/gregorcli/main.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+)
+
+func init() {
+	flag.StringVar(&tok, "auth-tok", "", "Authenitcation token")
+	flag.StringVar(&addr, "remote", "", "Remote address to test")
+}
+
+func main() {
+	flag.Parse()
+	conn, err := tls.Dial("tcp", addr, &tls.Config{InsecureSkipVerify: true})
```
Comment (maxtaco) [inline test/gregorcli/main.go]: can this also work for non-TLS? Might help in testing to see if plaintext TCP works but not TLS...

**Comment by maxtaco** [comment]
one question..

File: test/gregorcli/main.go (hunk 2 above)
Comment (jacobhaven) [inline test/gregorcli/main.go]: We don't have plain TCP deployed.

Comment (maxtaco) [inline test/gregorcli/main.go]: This is good for now so ship. In the future I want to see: (1) ability to
connect over plaintext tcp and also (2) the ability to pass CAs in for
testing.  For testing in various environments

Comment (jacobhaven) [inline test/gregorcli/main.go]: Well, this doesn't work yet. The auth token isn't verifying in `authd`.

Comment (maxtaco) [inline test/gregorcli/main.go]: Any more data?  What does the data look like?  What's the error message on authd?

Comment (jacobhaven) [inline test/gregorcli/main.go]: both **done**

**Comment by jacobhaven** [comment]
This can now be run as `gregorcli -auth-tok "$AUTHTOK" -remote fmprpc+tls://gregord.dev.keybase.io:443 -ca /path/to/dev_root_ca_cert.pem`
Returning:
```

[D] {tcp://52.72.123.193:443} New connection
[D] {tcp://52.72.123.193:443} call(0): method=gregor.1.auth.authenticateSessionToken; err=null; arg=[{"session":"$AUTHTOK"}];
[D] {tcp://52.72.123.193:443} reply(0): method=gregor.1.auth.authenticateSessionToken; err=null; res={"uid":"JpS0RkKgaD0yGtP4165fAA==","sid":"a652ed02c379412eb6ee4b6a390af902"};
[P] {tcp://52.72.123.193:443} call gregor.1.auth.authenticateSessionToken ran in 14ms
[D] {tcp://52.72.123.193:443} call(1): method=gregor.1.incoming.ping; err=null; arg=[{}];
[D] {tcp://52.72.123.193:443} reply(1): method=gregor.1.incoming.ping; err=null; res="pong";
[P] {tcp://52.72.123.193:443} call gregor.1.incoming.ping ran in 2ms
[D] {tcp://52.72.123.193:443} call(2): method=gregor.1.incoming.consumeMessage; err=null; arg=[{"m":{"ibm":{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"QgRLbeAXkFB0niRLGqyc2A==","ctime":1461264725459,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"ZmFrZSBib2R5"}}}}}];
[D] {tcp://52.72.123.193:443} serve(0): method=gregor.1.outgoing.broadcastMessage; err=null; arg=[{"m":{"ibm":{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"QgRLbeAXkFB0niRLGqyc2A==","ctime":1461264725459,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"ZmFrZSBib2R5"}}}}}];
received message: gregor1.DeviceID{0x66, 0x61, 0x6b, 0x65, 0x64, 0x65, 0x76, 0x69, 0x64, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37}
[P] {tcp://52.72.123.193:443} serve gregor.1.outgoing.broadcastMessage ran in 0ms
[D] {tcp://52.72.123.193:443} reply(0): method=gregor.1.outgoing.broadcastMessage; err=null; res=null;
[D] {tcp://52.72.123.193:443} reply(2): method=gregor.1.incoming.consumeMessage; err=null; res=null;
[P] {tcp://52.72.123.193:443} call gregor.1.incoming.consumeMessage ran in 26ms
[D] {tcp://52.72.123.193:443} call(3): method=gregor.1.incoming.sync; err=null; arg=[{"uid":"JpS0RkKgaD0yGtP4165fAA==","deviceid":"ZmFrZWRldmlkMTIzNDU2Nw==","ctime":0}];
[D] {tcp://52.72.123.193:443} reply(3): method=gregor.1.incoming.sync; err=null; res={"msgs":[{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"ZmFrZW1zZ2lkMTIzNDU2Nw==","ctime":1461207623299,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"DQBeAAAMACAg"}}},{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"Xtn66JxdMs5DtakW6QJHng==","ctime":1461208009157,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"DQAgIGRiM2Zk"}}},{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"Q/AtulyTkATnuJs4lPGxRg==","ctime":1461253921896,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"MjI3NmVjNDU1"}}},{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"WwswOzIahgaFVsa1qPKOAw==","ctime":1461254600917,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"ZmY2ZWFkOWZk"}}},{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"2z/d132/Gfp9mbj4CaGSNg==","ctime":1461254689464,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"NDQzZWE5MTAx"}}},{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"QiduxFWyB4o/nxQdugoQnw==","ctime":1461254911036,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"NzQ5ZTI0NGIx"}}},{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"ruU+/26tn9eBvf/nrkOIHw==","ctime":1461254945519,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"ZmFrZSBib2R5"}}},{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"jyXqFHNkQ+qRAYrouXwROA==","ctime":1461259937336,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"ZmFrZSBib2R5"}}},{"stateUpdate":{"md":{"uid":"JpS0RkKgaD0yGtP4165fAA==","msgID":"QgRLbeAXkFB0niRLGqyc2A==","ctime":1461264725459,"deviceID":"ZmFrZWRldmlkMTIzNDU2Nw==","inBandMsgType":1},"creation":{"category":"foo","dtime":{"time":0,"offset":0},"notifyTimes":null,"body":"ZmFrZSBib2R5"}}}],"hash":"5yQQ9kx7BYo+iUHTPmU7VGY87lVnBJmuGvYIuLHeHpA="};
[P] {tcp://52.72.123.193:443} call gregor.1.incoming.sync ran in 7ms
9 messages synced
```

**Comment by maxtaco** [comment]
Very nice

**Comment by maxtaco** [comment]
👍

**Pull request merged**